tool/tbot/botfs/fs_linux.go

```diff
@@ -0,0 +1,454 @@
+			return nil, trace.Wrap(err)
+		}
+	case SymlinksInsecure:
+		file, err = openStandard(path)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+	default:
+		return nil, trace.BadParameter("invalid symlinks mode %q", symlinksMode)
+	}
+
+	return file, nil
+}
+
+// createStandard creates an empty file or directory at the given path while
+// attempting to prevent symlink attacks.
+func createSecure(path string, isDir bool) error {
+	if isDir {
+		// We can't specify RESOLVE_NO_SYMLINKS for mkdir. This isn't the end
+		// of the world, though: if an attacker attempts a symlink attack we'll
+		// just open the correct file for read/write later (and error when it
+		// doesn't exist).
+		if err := os.Mkdir(path, DefaultDirMode); err != nil {
+			return trace.Wrap(err)
+		}
+	} else {
+		f, err := openSecure(path)
+		if err == unix.ENOSYS {
+			// bubble up the original error for comparison
+			return err
+		} else if err != nil {
+			return trace.Wrap(err)
+		}
+
+		// No writing to do, just close it.
+		f.Close()
+	}
+
+	return nil
+}
+
+// Create attempts to create the given file or directory with the given
+// symlinks mode.
+func Create(path string, isDir bool, symlinksMode SymlinksMode) error {
+	// Implementation note: paranoid file _creation_ is only really useful for
+	// providing an early warning if openat2() / ACLs are unsupported on the
+	// host system, as it will catch compatibility issues during `tbot init`.
+	// Read() and Write() with Symlinks(Try)Secure are the codepaths that
+	// actually prevents symlink attacks.
+
+	switch symlinksMode {
+	case SymlinksSecure:
+		if err := createSecure(path, isDir); err != nil {
+			if err == unix.ENOSYS {
+				return trace.Errorf("createSecure(%q) failed due to missing "+
+					"syscall; `symlinks: insecure` may be required for this "+
+					"system", path)
+			}
+
+			return trace.Wrap(err)
+		}
+	case SymlinksTrySecure:
+		err := createSecure(path, isDir)
+		if err == nil {
+			// All good, move on.
+			return nil
+		}
+
+		if err != unix.ENOSYS {
+			// Something else went wrong, fail.
+			return trace.Wrap(err)
+		}
+
+		// TODO: this will be very noisy on older systems. Maybe flip a global
+		// or something to only log once?
+		log.Warnf("Failed to create %q securely due to missing syscall; "+
+			"falling back to regular file creation. Set `symlinks: "+
+			"insecure` on this destination to disable this warning.", path)
+
+		return trace.Wrap(createStandard(path, isDir))
+	case SymlinksInsecure:
+		return trace.Wrap(createStandard(path, isDir))
+	}
+
+	return nil
+}
+
+// Read reads the contents of the given file into memory.
+func Read(path string, symlinksMode SymlinksMode) ([]byte, error) {
+	file, err := openSymlinksMode(path, symlinksMode)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	defer file.Close()
+
+	// Blatantly stolen from go's os/file.go since there's no stdlib function
+	// for reading everything from an _already open_ file, sigh.
```

Comment:
What about `io.ReadFull(f)`?

Reply:
I think `io.ReadAll()` is even better. Honestly perplexed how my googling failed to find that one, I guess I'm losing my touch.